src/utils.js

```diff
@@ -138,52 +139,66 @@ export function decodeStoryArgs(value) {
+    try {
+      // provide discovery options that may impact how the page loads
+      let page = yield percy.browser.page({
+        networkIdleTimeout: percy.config.discovery.networkIdleTimeout,
+        requestHeaders: getAuthHeaders(percy.config.discovery),
+        captureMockedServiceWorker: percy.config.discovery.captureMockedServiceWorker,
+        userAgent: percy.config.discovery.userAgent
+      });
+
+      // patch eval to include storybook specific helpers in the local scope
+      page.eval = (fn, ...args) => page.constructor.prototype.eval.call(page, (
+        typeof fn === 'string' ? fn : [
+          'function withPercyStorybookHelpers() {',
+          `  const VAL_REG = ${VAL_REG};`,
+          `  const NUM_REG = ${NUM_REG};`,
+          `  const HEX_REG = ${HEX_REG};`,
+          `  const COL_REG = ${COL_REG};`,
+          `  const VALID_REG = ${VALID_REG};`,
+          `  return (${fn})(...arguments);`,
+          `  ${isPlainObject}`,
+          `  ${validateStoryArgs}`,
+          `  ${encodeStoryArgs}`,
+          `  ${decodeStoryArgs}`,
+          '}'
+        ].join('\n')
+      ), ...args);
+
+      try {
+        yield page.goto(url);
+        return yield* yieldTo(callback(page));
+      } catch (error) {
+        // if the page crashed and retry returns truthy, try again
+        if (error.message?.includes('crashed') && retry?.()) {
+          return yield* withPage(...arguments);
+        }
 
-  // patch eval to include storybook specific helpers in the local scope
-  page.eval = (fn, ...args) => page.constructor.prototype.eval.call(page, (
-    typeof fn === 'string' ? fn : [
-      'function withPercyStorybookHelpers() {',
-      `  const VAL_REG = ${VAL_REG};`,
-      `  const NUM_REG = ${NUM_REG};`,
-      `  const HEX_REG = ${HEX_REG};`,
-      `  const COL_REG = ${COL_REG};`,
-      `  const VALID_REG = ${VALID_REG};`,
-      `  return (${fn})(...arguments);`,
-      `  ${isPlainObject}`,
-      `  ${validateStoryArgs}`,
-      `  ${encodeStoryArgs}`,
-      `  ${decodeStoryArgs}`,
-      '}'
-    ].join('\n')
-  ), ...args);
-
-  try {
-    yield page.goto(url);
-    return yield* yieldTo(callback(page));
-  } catch (error) {
-    // if the page crashed and retry returns truthy, try again
-    if (error.message?.includes('crashed') && retry?.()) {
-      return yield* withPage(...arguments);
+        /* istanbul ignore next: purposefully not handling real errors */
+        throw (typeof error !== 'string' ? error : new Error(error.replace(
+          // strip generic error names and confusing stack traces
+          /^Error:\s((.+?)\n\s+at\s.+)$/s,
+          // keep the stack trace if the error came from a client script
+          /\n\s+at\s.+?\(https?:/.test(error) ? '$1' : '$2'
+        )));
+      } finally {
+        // always clean up and close the page
+        await page?.close();
+      }
+    } catch (error) {
+      attempt++;
+      let shouldRetry = process.env.ENABLE_RETRY === 'true';
```

Comment:
Should we name something specific ENABL_RETRY_ON_STORY_ERROR something like this?, not sure we might add these kind of retry mechanism in CLI or somewhere else in future, so might confuse that time.